Extensions/DialogueTree/dialoguetools.js

```diff
@@ -0,0 +1,583 @@
+
+  this.optionsCount = 0;
+  this.selectedOption = -1;
+  this.selectedOptionUpdated = false;
+
+  if (gdjs.dialogueTree._lineTypeIsText()) {
+    if (
+      this.dialogueDataType === "options" ||
+      this.dialogueDataType === "text" ||
+      !this.dialogueDataType
+    ) {
+      this.clipTextEnd = 0;
+      this.dialogueText = this.dialogueData.text;
+      this.commandCalls = [];
+    } else {
+      this.dialogueText += this.dialogueData.text;
+    }
+
+    this.dialogueDataType = "text";
+    this.dialogueBranchTags = this.dialogueData.data.tags;
+    this.dialogueBranchTitle = this.dialogueData.data.title;
+    this.dialogueBranchBody = this.dialogueData.data.body;
+    this.dialogueData = this.dialogue.next().value;
+  } else if (gdjs.dialogueTree._lineTypeIsOptions()) {
+    this.dialogueDataType = "options";
+    this.optionsCount = this.dialogueData.options.length;
+    this.options = this.dialogueData.options;
+    this.selectedOptionUpdated = true;
+  } else if (gdjs.dialogueTree._lineTypeIsCommand()) {
+    this.dialogueDataType = "command";
+
+    var command = this.dialogueData.text.split(" ");
+    // if last command was to wait, increase time by one
+    var offsetTime =
+      this.commandCalls.length &&
+      this.commandCalls[this.commandCalls.length - 1].cmd === "wait"
+        ? 1
+        : 0;
+    this.commandCalls.push({
+      cmd: command[0],
+      params: command,
+      time: this.dialogueText.length + offsetTime
+    });
+    this.dialogueData = this.dialogue.next().value;
+    gdjs.dialogueTree.advanceDialogue();
+  } else {
+    this.dialogueDataType = "unknown";
+  }
+
+  if (gdjs.dialogueTree._lineTypeIsCommand()) {
+    this.dialogueDataType = "command";
+    gdjs.dialogueTree.advanceDialogue();
+  }
+
+  // dialogue has finished
+  if (!this.dialogueData) {
+    this.dialogueIsRunning = false;
+  }
+};
+
+/**
+ * use this to get the current Dialogue Tree branch title
+ */
+gdjs.dialogueTree.getBranchTitle = function() {
+  if (this.dialogueIsRunning) {
+    return this.dialogueBranchTitle;
+  }
+  return "";
+};
+
+/**
+ * use this to check if the currently parsed Dialogue branch title is a  query.
+ * @param {string} title The Dialogue Branch name you want to check for.
+ */
+gdjs.dialogueTree.branchTitleIs = function(title) {
+  if (this.dialogueIsRunning) {
+    return this.dialogueBranchTitle === title;
+  }
+  return false;
+};
+
+/**
+ * use this to get all the branch tags from the current Dialogue branch as a string. Useful for debugging.
+ */
+gdjs.dialogueTree.getBranchTags = function() {
+  if (this.dialogueIsRunning) {
+    return this.dialogueBranchTags.join(",");
+  }
+  return "";
+};
+
+/**
+ * use this to get one of the current Dialogue branch tags via index.
+ * @param {number} index The index of the Dialogue Branch tag you want to get.
+ */
+gdjs.dialogueTree.getBranchTag = function(index) {
+  if (this.dialogueIsRunning) {
+    if (index > this.dialogueBranchTags.length - 1)
```

Comment:
Is it possible for this array to be empty? If yes you should guard against this and return an empty string maybe?